torch/_inductor/codegen/cpp_bmm_template.py

```diff
@@ -0,0 +1,162 @@
+            {%- if is_dynamic_M %},
+            {{kernel.size(GemmOut, -2)}}
+            {% endif %}
+            
+        );
+    }
+    for (int64_t b_start = B_single_thread_block; b_start < B; ++b_start) {
+        blocked_mm(
+            &{{kernel.index(X, ["b_start", 0, 0])}},
+            &{{kernel.index(W, ["b_start", 0, 0])}},
+            &{{kernel.index(Y, ["b_start", 0, 0])}}
+            {%- if is_dynamic_M %},
+            {{kernel.size(GemmOut, -2)}}
+            {% endif %}
+        );
+    }
+}
+"""
+
+
+class CppBmmTemplate(CppPackedGemmTemplate):
+    def __init__(
+        self,
+        input_nodes,
+        layout: ir.Layout,
+        num_threads: int,
+        register_blocking: GemmBlocking,
+        beta=1,
+        alpha=1,
+        has_bias=False,
+        epilogue_creator: Optional[Callable[[ir.Buffer], ir.Pointwise]] = None,
+        name="bmm"
+    ):
+        super().__init__(
+            input_nodes,
+            layout,
+            num_threads,
+            register_blocking,
+            beta=beta,
+            alpha=alpha,
+            has_bias=has_bias,
+            epilogue_creator=epilogue_creator,
+            name=name
+        )
+        self.should_pack_weights = False
+
+    @staticmethod
+    def add_choices(
+        choices,
+        layout,
+        input_nodes,
+        beta=1,
+        alpha=1,
+        has_bias=False,
+        trans_w=False,
+        input_indices=None,
+        should_pack_weights=False,
+        epilogue_creator: Optional[Callable[[ir.Buffer], ir.Pointwise]] = None,
+    ):
+        options = super(CppBmmTemplate, CppBmmTemplate)._get_params_for_choices(
+            layout=layout,
+            input_nodes=input_nodes,
+            beta=beta,
+            alpha=alpha,
+            has_bias=has_bias,
+            trans_w=trans_w,
+            input_indices=input_indices,
+            should_pack_weights=should_pack_weights,
+            epilogue_creator=epilogue_creator
+        )
+        template = DataProcessorTemplateWrapper(
+            CppBmmTemplate,
+            **options
+        )
+        template.maybe_append_choice(choices)
+        return template
+
+    def render(  # type: ignore[override]
+        self,
+        kernel: CppTemplateKernel,
+        template_buffer_node: Optional[ir.CppTemplateBuffer] = None,
+        epilogue_nodes: Optional[List[ir.IRNode]] = None,
+        **kwargs,
+    ) -> str:
+        options = self.get_options(kernel, template_buffer_node, epilogue_nodes, **kwargs)
+        options['should_pack_weights'] = self.should_pack_weights
+        options['is_bmm'] = True
+
+        result = self._template_from_string(MICROKERNEL_DEF).render(**options)
+        result += self._template_from_string(BLOCKED_STUB+GEMM_TEMPLATE).render(**options)
+        self.thread_blocking.clear_cache(self)
+        self.cache_blocking.clear_cache(self)
+        tmp_num_threads = self.num_threads
+        self.num_threads = 1
```

Comment:
Perhaps it's better to factor out some context util to handle this thread_num overrides for gemm_template codegen and don't mess up with current states.

Reply:
I changed the caching to be stored along with the number of threads, and can be reset if number of threads changes.